### PR TITLE
Add golden tests for signing of transactions

### PR DESCRIPTION
**Pull request by Anviking**
# Issue Number

#95 

# Overview

- [x] I have added one simple golden test for transaction signing from the cardano-sl implementation
- [x] I have refactored and fixed the signing code according to this test

# Comments

- This is essentially only *one* golden test, and not particularly reassuring. Would be nice to generate a couple 100?

### CODE REVIEW

File: src/Cardano/Wallet/Primitive/Types.hs [hunk 1]
```diff
@@ -396,7 +396,7 @@ newtype SignedTx = SignedTx { signedTx :: ByteString }
     deriving (Show, Eq, Generic)
 
 data TxWitness
-    = PublicKeyWitness ByteString
+    = PublicKeyWitness ByteString (Hash "signature")
```
Comment (Anviking) [inline src/Cardano/Wallet/Primitive/Types.hs:399]: `Key 'AddressK XPub` would feel better instead of `ByteString`, will do in separate pr because of import cycles

File: test/unit/Cardano/Wallet/Primitive/SigningSpec.hs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        it "slAddress xpub == walletAddress xpub" $ do
+            bin <- readFixture "sl-addr.bin"
+            let seed = Passphrase $ BA.convert $ BS.pack $ replicate 128 0
+            let pass = Passphrase mempty
+            let addrXPrv = unsafeGenerateKeyFromSeed (seed, mempty) pass
+            let (Address addr) = keyToAddress $ publicKey addrXPrv
+            addr `shouldBe` bin
+
+
+        it "decodeSignedTx slBinary == walletTx" $ do
```
Comment (Anviking) [inline test/unit/Cardano/Wallet/Primitive/SigningSpec.hs]: I opted to test both using `decodeSignedTx` and `encodeSignedTx`. Should add a roundtrip property regardless (I think it doesn't exist. We only test the roundtrip on txs from the BinarySpec blocks.)

File: src/Cardano/Wallet/Binary.hs [hunk 3]
```diff
@@ -431,11 +431,21 @@ decodeTxWitness = do
...
-        0 -> CBOR.decodeTag *> (PublicKeyWitness <$> CBOR.decodeBytes)
+        0 -> do
+            _ <- CBOR.decodeTag
+            bs <- CBOR.decodeBytes
+            case CBOR.deserialiseFromBytes decodePKWitness (BL.fromStrict bs) of
+                Left err -> fail $ show err
+                Right (_, input) -> return input
         1 -> CBOR.decodeTag *> (ScriptWitness <$> CBOR.decodeBytes)
         2 -> CBOR.decodeTag *> (RedeemWitness <$> CBOR.decodeBytes)
         _ -> fail
```
Comment (Anviking) [inline src/Cardano/Wallet/Binary.hs:442]: As I had problems with it previously, I should check that this fail works, and doesn't produce ~infinite output.

File: src/Cardano/Wallet/Primitive/Types.hs (hunk 1 above)
Comment (KtorZ) [reply src/Cardano/Wallet/Primitive/Types.hs:399]: `XPub` would probably be sufficient here.

File: test/unit/Cardano/Wallet/Primitive/SigningSpec.hs [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        h = Hash "\bztb\249\223n5\189A\239Y&\243tgv\253\223\140\222i\236O\221\&8\205\218>\234\240\238"
+
+        ownedIns = (TxIn h 0, TxOut addr (Coin 1) )
+        outs = [TxOut addr (Coin 1)]
+
+        m = (Map.singleton addr addrXPrv) :: Map Address (Key 'AddressK XPrv)
+
+        fromR = either undefined Prelude.id
+    in
+        fromR $ mkStdTx m (error "no root key", pass) [ownedIns] outs
```
Comment (KtorZ) [inline test/unit/Cardano/Wallet/Primitive/SigningSpec.hs]: Similarly for what we did for `Yoroi` golden tests in the `AddressDerivationSpec` module, it would be better to have a `goldenSignedTx` utility to generate and compare a transaction binary format from a list of `(Address, Key 'AddressK XPub)`

File: test/unit/Cardano/Wallet/Primitive/SigningSpec.hs [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    = BS.readFile
+        $ "test/data/Cardano/Wallet/Primitive/Signing/"
+          ++ T.unpack (toText network') ++ "/" ++ name
+  where
+    network' =
+        case network of
+            Mainnet -> Mainnet
+            Staging -> Mainnet
+            Testnet -> Testnet
+            Local -> Testnet
```
Comment (KtorZ) [inline test/unit/Cardano/Wallet/Primitive/SigningSpec.hs]: If I recall correctly, the witness includes the protocol magic too, no ? So all tx should be slightly different even between staging and mainnet. Also, for consistency, we're probably better off with having the same file for local and testnet (not sure there's actually any value checking this for the local network...)

**Review by KtorZ — APPROVED** [review]
